File: day14/.js
/**
 * @param {object} workshop - A representation of the workshop
 * @param {string|number|boolean} gift - The gift to find
 * @returns {number[]} The path to the gift
 */
function findGiftPath(workshop, gift) {
  for (const key in workshop){
    if (workshop[key] == gift){
      return [key]
    }
    if (typeof workshop[key] == "object" && workshop[key] != null){
      const path = findGiftPath(workshop[key], gift)
      if (path.length > 0){
        return [key, ...path]
      }
    }
  }
  return [];
}

File: day14/.py
def find_gift_path(workshop: dict, gift: str | int | bool) -> list[str]:
  for key in workshop:
    if workshop[key] == gift:
      return [key]
    if isinstance(workshop[key], dict) and workshop[key] is not None:
      path = find_gift_path(workshop[key], gift)
      if len(path) > 0:
        return [key] + path
  return [];
